enarx-keep-sgx-shim/src/start.S

```diff
@@ -56,26 +117,112 @@ shp:                            # The bottom of the secure heap
-    pop     %rsp                                    # Restore old stack
-    mov     $4,                     %rax            # EEXIT
-    enclu                                           # ENCLU[EEXIT]
+    xchg    %rbx,                   %rcx            # Swap TCS and next instruction.
+    cmp     $0,                     %rax            # If CSSA > 0...
+    jne     .Levent                                 # ... restore stack from AEX[CSSA-1].
+
+# CSSA == 0
+    rdrand  %r11                                    # Get a random number
+    and     $0xff0,                 %r11            # Mask to an aligned random offset
+    lea     stk(%rip),              %r10            # Get the stack pointer
+    sub     %r11,                   %r10            # Subtract random offset
+
+    xor     %r11,                   %r11            # %r11 = NULL
+    xor     %rax,                   %rax            # %rax = NULL
+    jmp     .Lstack
+
+# CSSA != 0
+.Levent:
+    dec     %rax                                    # %rax = CSSA - 1
+    shl     $12,                    %rax            # %rax = (CSSA - 1) * 4096
+    movabs  $aex,                   %r11            # %r11 = &aex
+    add     %rax,                   %r11            # %r11 = &aex[CSSA - 1]
+
+    mov     RSP(%r11),              %r10            # %r10 = aex[CSSA - 1].gpr.rsp
+    sub     $128,                   %r10            # Skip the red zone
+    and     $~0xf,                  %r10            # Align
+
+    mov     SRSP(%r11),             %rax            # %rax = syscall return stack pointer
+    jmp     .Lstack
+
+# %rax = NULL or syscall return stack pointer
+# %rbx = next non-enclave instruction
+# %rcx = &TCS
+# %r10 = trusted stack pointer
+# %r11 = NULL or &aex[CSSA - 1]
+# %rsp = untrusted stack pointer
+.Lstack:
+    xchg    %r10,                   %rsp            # Swap to trusted stack
+    pushq   $0                                      # Align stack
+    push    %r10                                    # Save untrusted %rsp
+    savep                                           # Save untrusted preserved registers
+
+    cmp     $0,                     %rax            # If we are returning from a syscall...
+    jne     .Lsyscall                               # ... finish the job.
+
+    push    %rsp                                    # Argument for enter()
+    push    %r11                                    # Argument for enter()
+
+    zerop                                           # Clear preserved registers
+    zerot                                           # Clear temporary registers
+    zerof   %r11                                    # Clear CPU flags
+
+    # void enter(rdi, rsi, rdx, tcs, r8, r9, &aex[CSSA-1], ctx);
+    call    enter                                   # Call enter()
+    add     $16,                    %rsp            # Remove parameters from stack
+
+    # Prepare CPU context for exit
```

Comment:
It seems this is not sufficient? You don't clear all general purpose registers like R15 etc? AFAIK no ABI will guarantee they don't contain secrets and you should generally clear _all_ CPU state before doing EEXIT..

Also, crucially, your entry/exit code does _not_ consider CPU extended state (including XMM/YMM/etc registers used by for instance AES-NI)? You should use `XSAVE` and friends for this -- see eg Intel SGX SDK entry/exit code.

Reply:
> It seems this is not sufficient? You don't clear all general purpose registers like R15 etc? AFAIK no ABI will guarantee they don't contain secrets and you should generally clear _all_ CPU state before doing EEXIT..

The preserved registers are loaded below on line 182. The only GPR register we don't clear is `%rax`, which is set to a constant for `EEXIT`.

> Also, crucially, your entry/exit code does _not_ consider CPU extended state (including XMM/YMM/etc registers used by for instance AES-NI)? You should use `XSAVE` and friends for this -- see eg Intel SGX SDK entry/exit code.

This was intentionally left out to reduce the review size on this already difficult-to-review PR. It is also something that can be done from Rust. We have an issue for it: https://github.com/enarx/enarx/issues/334